### ISSUE TITLE
Direct dependency to React?

### ISSUE THREAD

**Issue by screendriver** [garris/BackstopJS#695]
Hi,
why did you added React and ReactDOM as a [dependency](https://github.com/garris/BackstopJS/blob/master/package.json#L78) to BackstopJS? Since that point I can't use BackstopJS anymore because my project depends on React 16.1. and BackstopJS relies on React 15. When I run my tests with [Jest](http://facebook.github.io/jest/) they crash because of different versions of React. With BackstopJS `3.0.39` everything worked fine in the past.

**Comment by mantovanig** [comment]
Hi @screendriver ,

yes the issue is that react dependencies should be declared as "Peer Dependencies" in package.json
https://nodejs.org/en/blog/npm/peer-dependencies/

We will do it as soon as possible.

Thanks

**Comment by screendriver** [comment]
Hi @mantovanig,

ok, cool. But why do you have a dependency to React in general? If you made it as a `peerDependency` you are assuming / forcing the consuming project to use React. What if you want to use BackstopJS but don't have nothing to do with React? For example if your project uses Angular or Vue. These projects are now forced to install React just for BackstopJS? 🤔

**Comment by matthiaskomarek** [comment]
Hey @screendriver,

the html report page, which gets generated to show your test results, uses React. Thats the reason why `BackstopJS` has this dependency.

**Comment by mantovanig** [comment]
It's right @matthiaskomarek ,

but when you use backstopJs inside a react project there are react dependencies conflicts, because in backstopjs react is not declared as peerDependencies.

On the other hand if we declared react as peerDependencies, react become a requirement for every project.

It would serve something like optionalPeerDependencies, but it doesn't exist :)

I think about it

**Comment by matthiaskomarek** [comment]
@mantovanig i would like to have react as peerDep.

Maybe it would be good to upgrade to React 16.x and also make it a peerDep. What do you think?
I am curious what @garris thinks about this.

**Comment by mantovanig** [comment]
Yes @matthiaskomarek , probably this is the best solution at the moment.

If it's right even for @garris, I can do this fix tomorrow.

**Comment by garris** [comment]
That ^^^ article is fantastic...

> The peer dependency idea has been kicked around for literally years. After volunteering to get this done "over the weekend" nine months ago, I finally found a free weekend, and now peer dependencies are in npm!

**Comment by garris** [comment]
I don’t have an opinion yet — I still need to understand the problem better.  But, everyone will always have a different version of react so we will never be able to make everyone happy.  I am fine to just check a static version of react library into our repo if that makes this problem go away.

**Comment by garris** [comment]
In backstop — react is bundled by webpack during development.  Afterward, it is not used.   It seems unnecessary to cause global conflicts for a library we use locally this way.

Questions:

Do we gain anything by calling it a devdependency?

Also, why should backstop cause an issue for other apps?  If backstop were listed after react in a package.json wouldn’t its version of react be included locally (and not be used by the parent app?)   

Referencing by order method should totally work for us since react is only referenced at build during backstop dev.

Also, I  am probably wrong about something.  

Comments?

**Comment by longstaff** [comment]
Added pull request for this issue here: #748

**Comment by garris** [comment]
@longstaff    I just discovered that #748 breaks local development on the web report.   I decided to revert because it seems that this solution needs just a little bit more work.   Would love it if you or someone else could take another stab at it.

**Comment by longstaff** [comment]
@garris peerDependency is expected to be installed separately (you should have gotten an error in your npm install that you have a missing peerDependency on react) which will mean that for people using your plugin their local system should have a dependency on react. When you are runnning local dev you would have to install it separately, or we can add it to devDependencies as well.

**Comment by visionmonster** [comment]
I think you would want all of the react / webpack code as a devDependencies sections. I think this would hold for any dependency that only applies to generating the compare/output code.

from the npm docs:
https://docs.npmjs.com/files/package.json#devdependencies

> If someone is planning on downloading and using your module in their program, then they probably don't want or need to download and build the external test or documentation framework that you use
In this case, it's best to map these additional items in a devDependencies object.

It's fine to have it as a dependency if it's an application, but I think the trouble comes from using it as a Module especially with react.

fantastic library by the way.
